src/component/cyfs-stack/src/stack/group_non_driver.rs

```diff
@@ -0,0 +1,265 @@
+    ) -> BuckyResult<Option<NONObjectInfo>> {
+        let obj_type_code = obj.object_id.obj_type_code();
+        let obj_type = obj.object.as_ref().map(|obj| obj.obj_type());
+
+        let req_path = RequestGlobalStatePath::new(Some(dec_id.clone()), Some("group/inner-cmd"));
+
+        self.non_service
+            .post_object(NONPostObjectInputRequest {
+                common: NONInputRequestCommon {
+                    req_path: Some(req_path.format_string()),
+                    source: RequestSourceInfo {
+                        protocol: RequestProtocol::DataBdt,
+                        zone: DeviceZoneInfo {
+                            device: None,
+                            zone: None,
+                            zone_category: DeviceZoneCategory::CurrentZone,
+                        },
+                        dec: dec_id.clone(),
+                        verified: None,
+                    },
+
+                    level: NONAPILevel::Router, // to.map_or(NONAPILevel::NOC, |_| NONAPILevel::Router),
+
+                    target: to.cloned(),
+                    flags: 0,
+                },
+                object: obj,
+            })
+            .await
+            .map(|resp| resp.object)
+            .map_err(|err| {
+                log::warn!(
+                    "group post object(type={:?}/{:?}) to {:?} failed {:?}",
+                    obj_type_code,
+                    obj_type,
+                    to,
+                    err
+                );
+                err
+            })
+    }
+}
+
+#[async_trait::async_trait]
+impl cyfs_group::NONDriver for GroupNONDriver {
+    async fn get_object(
+        &self,
+        dec_id: &ObjectId,
+        object_id: &ObjectId,
+        from: Option<&ObjectId>,
+    ) -> BuckyResult<NONObjectInfo> {
+        let fut1 = match futures::future::select(
+            self.get_object_impl(dec_id, object_id, from).boxed(),
+            async_std::future::timeout(TIMEOUT_HALF, futures::future::pending::<()>()).boxed(),
+        )
+        .await
+        {
+            futures::future::Either::Left((ret, _)) => return ret,
+            futures::future::Either::Right((_, fut)) => fut,
+        };
+
+        log::warn!(
+            "group get object timeout (type={:?}) from {:?}, local: {:?}",
+            object_id.obj_type_code(),
+            from,
+            self.local_device_id
+        );
+
+        match futures::future::select(
+            self.get_object_impl(dec_id, object_id, from).boxed(),
+            async_std::future::timeout(TIMEOUT_HALF, fut1).boxed(),
+        )
+        .await
+        {
+            futures::future::Either::Left((ret, _)) => ret,
+            futures::future::Either::Right((ret, _)) => ret.map_or(
+                Err(BuckyError::new(BuckyErrorCode::Timeout, "timeout")),
+                |ret| ret,
+            ),
+        }
+    }
+
+    async fn put_object(&self, dec_id: &ObjectId, obj: NONObjectInfo) -> BuckyResult<()> {
```

Comment:
Why not just use sync_std::future::timeout instead of relying on the select of futures, which is obviously more complicated?

Reply:
I want to hold the life of the future to continue. and start a new future to retry at the same time. maybe the first future is slow only.